### PR TITLE
PHP-1410: Add certificate expiration checks

### PR DESCRIPTION
**Pull request by bjori**
This builds on PHP-1407 and PHP-1406

### CODE REVIEW

File: io_stream.c [hunk 1]
```diff
@@ -141,6 +169,53 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
...
+					mongo_manager_log(manager, MLOG_CON, MLOG_INFO, "Could not capture certificate of %s:%d", server->host, server->port);
+					php_stream_close(stream);
+					return NULL;
+				}
+
+#if PHP_VERSION_ID < 50600
+				/* This option is available since PHP 5.6.0 */
+				if (php_stream_context_get_option(options->ctx, "ssl", "verify_peer_name", &verify_peer_name) == SUCCESS && zend_is_true(*verify_peer_name)) {
+					if (php_mongo_verify_hostname(server, cert TSRMLS_CC) == FAILURE) {
+						*error_message = strdup("Hostname doesn't match");
```
Comment (derickr) [inline io_stream.c]: Shouldn't this have the expected, and the given hostname? And clarify it is about SSL peer verifrication?

File: io_stream.c [hunk 2]
```diff
@@ -141,6 +169,53 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
...
+				}
+#endif
+				if (php_stream_context_get_option(options->ctx, "ssl", "verify_expiry", &verify_expiry) == SUCCESS && zend_is_true(*verify_expiry)) {
+					time_t current = time(NULL);
+					time_t valid_from  = php_mongo_asn1_time_to_time_t(X509_get_notBefore(cert) TSRMLS_CC);
+					time_t valid_until = php_mongo_asn1_time_to_time_t(X509_get_notAfter(cert) TSRMLS_CC);
+
+					if (valid_from > current) {
+						mongo_manager_log(manager, MLOG_CON, MLOG_INFO, "Certificate is not valid yet on %s:%d", server->host, server->port);
+						php_stream_close(stream);
```
Comment (derickr) [inline io_stream.c]: Although you return NULL here, how does the calling code know what the reason is? (Without checking the log)

File: io_stream.c [hunk 3]
```diff
@@ -141,6 +169,53 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
...
+						return NULL;
+					}
+					if (current > valid_until) {
+						mongo_manager_log(manager, MLOG_CON, MLOG_INFO, "Certificate has expired on %s:%d", server->host, server->port);
+						php_stream_close(stream);
```
Comment (derickr) [inline io_stream.c]: Same here

File: io_stream.c [hunk 4]
```diff
@@ -141,6 +169,53 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
...
+
+#if PHP_VERSION_ID < 50600
+				/* This option is available since PHP 5.6.0 */
+				if (php_stream_context_get_option(options->ctx, "ssl", "verify_peer_name", &verify_peer_name) == SUCCESS && zend_is_true(*verify_peer_name)) {
+					if (php_mongo_verify_hostname(server, cert TSRMLS_CC) == FAILURE) {
+						*error_message = strdup("Hostname doesn't match");
+						php_stream_close(stream);
+						return NULL;
+					}
+					mongo_manager_log(manager, MLOG_CON, MLOG_INFO, "stream_connect: Valid peer name for %s:%d", server->host, server->port);
```
Comment (derickr) [inline io_stream.c]: Wouldn't we warn in the log if there is *no* valid peer certificate as well? Just like with the error_message that you return in 194?

File: io_stream.c (hunk 1 above)
Comment (bjori) [inline io_stream.c]: There is no one expected hostname.. There could be lots of alternative subject names in the certificate, along with the common name.
We do issue mismatch warning including the common name, which gets turned into an exception.
You have to iterate over the $exception->getPrevious() to get all the error messages

File: io_stream.c (hunk 2 above)
Comment (bjori) [inline io_stream.c]: Your right, we need to populate the error_message

**Comment by bjori** [comment]
should be all fixed now

File: io_stream.c [hunk 5]
```diff
@@ -116,7 +148,12 @@ void* php_mongo_io_stream_connect(mongo_con_manager *manager, mongo_server_def *
 
 		zend_replace_error_handling(EH_THROW, mongo_ce_ConnectionException, &error_handler TSRMLS_CC);
 
+		/* PHP5.6.0 until 5.6.7 screwed things a bit, see https://bugs.php.net/bug.php?id=69195 */
```
Comment (derickr) [inline io_stream.c]: Space between PHP and 5.6.0 please

**Comment by derickr** [comment]
LGTM, but:

 - [ ] Merge #834 first, and then merge merge #836 separately for 1407, 1410, and the extra bits

**Pull request merged**